doc/developer/design/20230705_v2_txn_management.md

```diff
@@ -0,0 +1,428 @@
+loop {
+    if commit_at(write_ts, update_fn(s, read_ts)) {
+        break;
+    }
+    (read_ts, write_ts) = oracle.read_and_write_ts();
+}
+```
+
+- If `update_fn` can be computed incrementally, there are various tricks we can
+  play to avoid re-writing all the data to s3 on each retry.
+- On retry, there's a potential performance special case to check if any of the
+  inputs to `update_fn` have changed since the last read_ts and skip recomputing
+  update_fn if not.
+  - Note that with this optimization, this does exactly the same thing as
+    "write-only single-table txn".
+
+The rest of this document concerns the implementation of the durability
+component. Additional detail on the timestamp selection is left for a separate
+design document.
+
+## Detailed description
+
+See accompanying prototype in [#20954].
+
+[#20954]: https://github.com/MaterializeInc/materialize/pull/20954
+
+Efficient atomic multi-shard writes are accomplished through a new
+singleton-per-environment _txns shard_ that coordinates writes to a (potentially
+large) number of _data shards_. Data shards may be added to and removed from the
+set at any time.
+
+Benefits of these txns:
+- _Key idea_: A transactional write costs in proportion to the total size of
+  data written, and the number of data shards involved (plus one for the txns
+  shard).
+- _Key idea_: As time progresses, the upper of every data shard is logically
+  (but not physically) advanced en masse with a single write to the txns shard.
+  (Data writes may also be bundled into this, if desired.)
+- Transactions are write-only, but read-then-write transactions can be built on
+  top by using read and write timestamps selected to have no possible writes in
+  between (e.g. `write_ts/commit_ts = read_ts + 1`).
+- Transactions of any size are supported in bounded memory. This is done though
+  the usual persist mechanism of spilling to s3. These spilled batched are
+  efficiently re-timestamped when a commit must be retried at a higher
+  timestamp.
+- The data shards may be read independently of each other.
+- The persist "maintenance" work assigned on behalf of the committed txn is
+  (usually, see below) assigned to the txn committer.
+- It is possible to implement any of snapshot, serializable, or
+  strict-serializable isolation on top of this interface via write and read
+  timestamp selections.
+- It is possible to serialize and communicate an uncommitted `Txn` between
+  processes and also to merge uncommitted `Txn`s, if necessary (e.g.
+  consolidating all monitoring collections, statement logging, etc into the
+  periodic timestamp advancement).
+
+Restrictions:
+- Data shards must all use the same codecs for `K, V, T, D`. However, each data
+  shard may have a independent `K` and `V` schemas. The txns shard inherits the
+  `T` codec from the data shards (and uses its own `K, V, D` ones).
+- All txns are linearized through the txns shard, so there is some limit to
```

Comment:
Are they linearized or serialized through the txns shard? IMO, the txn logic gives us serializability and the way we pick timestamps ensures linearizability. Which together ensures we provide _strict serializability_.

Reply:
> IMO, the txn logic gives us serializability and the way we pick timestamps ensures linearizability.

I think technically all write only txns are linearized through the txns shard. If a write `B` starts after write `A` completes then it's guaranteed to have a larger timestamp, because of the upper advancing. I.e. the total order of writes matches the real time ordering.

If we're talking about read then write txns, then I don't think the txn logic even provides serializability, at best it provides read committed and monotonic writes. The timestamp selection is what ensures both serializability and linearizability.

Reply:
☝️ I had a response staged, but joe's is much better, so I deleted it :)

Change this to "All txn writes are linearized"